### ISSUE TITLE
`gf` support for environment variables

### ISSUE THREAD

**Issue by Kratacoa** [mawww/kakoune#4064]
### Feature
Support for environment variables such as `$HOME` when using `gf`.  
e.g. `~/repository/boop.md` works, however `$HOME/repository/boop.md` doesn't.

**Comment by krobelus** [comment]
It's probably best to write a mapping that uses the shell to expand variables.
Kakoune doesn't parse shell, and `gf` is quite a low-level command.

```
define-command goto-file-shellexpand %{
    evaluate-commands %sh{
        eval echo "edit ${kak_reg_dot}"
    }
}
```

I like to use a mapping that guesses filename boundaries before running `gf`.
It also detects line/column numbers - place your cursor one something like  `Makefile:5:2`, and this command will go to line 5, column to:

```
define-command goto-file-smart %{
    execute-keys %{<a-a><a-w>1s([^: ]+)(?::(\d+))?(?::(\d+))?<ret>) }
    execute-keys -with-hooks "gf%reg{2}g<a-h>h%reg{3}l"
}
```

(that doesn't do shell expansion yet)

**Comment by Kratacoa** [comment]
>  Kakoune doesn't parse shell, and gf is quite a low-level command.

Disclaimer: not a programmer, just someone who toys with Linux
Doesn't `~` imply an understanding of expansions at a low level?

**Comment by krobelus** [comment]
You're right, `~` is expanded as well - usually it's the shell that does that. I think Kakoune makes an exception for this one because it's simple and very common.

**Comment by Kratacoa** [comment]
What would be the implications of generalising this thing, given that `~` is already accepted (and I assume works via the same mechanics that `$HOME` interpretation would use)?

**Comment by krobelus** [comment]
Kakoune would have to make some decisions that are not completely obvious, for example
- if there are two adjacent spaces, like `a  b`, should we interpret that literally, or as `a b`?
- is it okay to expand command substitutions?
- should it use the environment variables from the server process, or from the active client?

These are all solvable to some extent, however, the biggest problem with this feature is that it's specific to shell syntax.
It won't satisfy users who use different file formats.
For example, I sometimes use Makefiles where a path might look like `$(HOME)/foo.md`. It's more scalable to let the user choose how they want this to work.

I suggest you just map `gf` to a simple wrapper like `goto-file-shellexpand` (I'll tweak that a bit, it has some issues).

**Comment by andreyorst** [comment]
@Kratacoa I'm using a custom function for this purpose for a quite a long time: [search-file](https://gitlab.com/andreyorst/dotfiles/-/blob/787828a0676fe4c35210dca778ff704133894812/.config/kak/commands.kak#L59-118). It is basically an emulation of Vim's behavior of `gf`, which is capable of smart recursive search under several paths. Paired with another function - [smart-select-file](https://gitlab.com/andreyorst/dotfiles/-/blob/787828a0676fe4c35210dca778ff704133894812/.config/kak/commands.kak#L11-46) and mapped to `gf` like [this](https://gitlab.com/andreyorst/dotfiles/-/blob/787828a0676fe4c35210dca778ff704133894812/.config/kak/mappings.kak#L39-40) I have never used inbuilt `gf` feature of Kakoune. See if this works for you?
